### PR TITLE
Scabbard receipt store integration

### PR DESCRIPTION
**Pull request by isabeltomb**
This PR adds the new receipt store trait implemented in Sawtooth-lib to scabbard. The new receipt store is added to scabbard along side the existing receipt store while it is still experimental. The new receipt store is behind the experimental feature "diesel-receipt-store".

The new receipt store trait is implemented for both PostgreSQL and SQLite backends, scabbard will use the same PostgreSQL or SQLite instance that is being used by Splinter for the receipt store.

To test run splinter database migrate with experimental features turned on you should see either of the following messages:
```
Running migrations against PostgreSQL database for receipt store: postgres://postgres:<pg-password>@localhost:5432/splinter
Successfully applied PostgreSQL migrations
```
or
```
Running migrations against SQLite database for receipt store: /data/splinter_state.db
Successfully applied SQLite migrations
```
then start up a splinter node with experimental features on, look at the transaction receipts database tables start a second splinter node, create a circuit and follow the steps to upload a smart contract and grant necessary permissions. 

The smart contract setup steps should create 4 transaction receipts which can be viewed in the transaction_receipt table of the database being used with the SQL statement `SELECT * FROM 'transaction_receipt';`
Starting a workload will submit additional transactions and add more transaction receipts to the tables.

### CODE REVIEW

File: splinterd/src/daemon.rs [hunk 1]
```diff
@@ -453,6 +455,15 @@ impl SplinterDaemon {
             }
         }
 
+        #[cfg(feature = "scabbard-receipt-store")]
+        run_receipt_store_migrations(&self.db_url.to_string()).map_err(|err| {
```
Comment (peterschwarz) [inline splinterd/src/daemon.rs:459]: This should be run with the `splinter database migrate` command, not here in the daemon

File: services/scabbard/libscabbard/src/service/state.rs [hunk 2]
```diff
@@ -162,6 +164,45 @@ impl ScabbardState {
         // initialize committed_batches metric
         counter!("splinter.scabbard.committed_batches", 0);
 
+        #[cfg(feature = "diesel-receipt-store")]
+        let receipt_store: Arc<RwLock<dyn ReceiptStore>> =
```
Comment (peterschwarz) [inline services/scabbard/libscabbard/src/service/state.rs:168]: Should this be an alternative to the existing lmdb files?

File: services/scabbard/libscabbard/src/service/state.rs [hunk 3]
```diff
@@ -666,18 +671,25 @@ enum EventQuery {
 /// An iterator that wraps the `TransactionReceiptStore` and returns `StateChangeEvent`s using an
 /// in-memory cache.
 pub struct Events {
+    // transaction_receipt_store is not used if diesel-receipt-store is enabled
+    #[allow(unused_variables)]
     transaction_receipt_store: Arc<RwLock<TransactionReceiptStore>>,
+    #[cfg(feature = "diesel-receipt-store")]
+    receipt_store: Arc<RwLock<dyn ReceiptStore>>,
```
Comment (peterschwarz) [inline services/scabbard/libscabbard/src/service/state.rs:678]: These are different enough that I would actually suggest two different implementations of the struct, one for when the feature is enabled and one for `not`.

**Review by peterschwarz — CHANGES_REQUESTED** [review]

File: splinterd/src/daemon.rs (hunk 1 above)
Comment (isabeltomb) [reply splinterd/src/daemon.rs]: Moved this functionality to be run with `splinter database migrate` if "scabbard-receipt-store" is enabled

File: services/scabbard/libscabbard/src/service/state.rs (hunk 2 above)
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/service/state.rs:164]: Modified ScabbardState to only create the lmdb receipt store if "diesel-receipt-store" is not enabled

File: services/scabbard/libscabbard/src/service/state.rs (hunk 3 above)
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/service/state.rs]: Updated

**Review by peterschwarz — APPROVED** [review]

**Review by agunde406 — APPROVED** [review]
Tested locally.

**Pull request merged**